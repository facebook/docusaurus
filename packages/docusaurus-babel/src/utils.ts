/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import fs from 'fs-extra';
import path from 'path';
import {BABEL_CONFIG_FILE_NAME} from '@docusaurus/utils';
import type {TransformOptions} from '@babel/core';

export async function getCustomBabelConfigFilePath(
  siteDir: string,
): Promise<string | undefined> {
  const customBabelConfigurationPath = path.join(
    siteDir,
    BABEL_CONFIG_FILE_NAME,
  );
  return (await fs.pathExists(customBabelConfigurationPath))
    ? customBabelConfigurationPath
    : undefined;
}

export function getBabelOptions({
  isServer,
  babelOptions,
}: {
  isServer?: boolean;
  // TODO Docusaurus v4 fix this
  //  weird to have getBabelOptions take a babelOptions param
  babelOptions?: TransformOptions | string;
} = {}): TransformOptions {
  const caller = {name: isServer ? 'server' : 'client'};
  if (typeof babelOptions === 'string') {
    return {
      babelrc: false,
      configFile: babelOptions,
      caller,
    };
  }
  return {
    ...(babelOptions ?? {
      presets: [require.resolve('@docusaurus/babel/preset')],
    }),
    babelrc: false,
    configFile: false,
    caller,
  };
}
