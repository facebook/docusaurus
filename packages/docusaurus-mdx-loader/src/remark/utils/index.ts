/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import escapeHtml from 'escape-html';
import type {Parent, Node} from 'unist';
import type {PhrasingContent, Heading} from 'mdast';
import type {
  MdxJsxAttribute,
  MdxJsxAttributeValueExpression,
  MdxJsxTextElement,
  // @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
} from 'mdast-util-mdx';

/**
 * Util to transform one node type to another node type
 * The input node is mutated in place
 * @param node the node to mutate
 * @param newNode what the original node should become become
 */
export function transformNode<NewNode extends Node>(
  node: Node,
  newNode: NewNode,
): NewNode {
  Object.keys(node).forEach((key) => {
    // @ts-expect-error: unsafe but ok
    delete node[key];
  });
  Object.keys(newNode).forEach((key) => {
    // @ts-expect-error: unsafe but ok
    node[key] = newNode[key];
  });
  return node as NewNode;
}

export function stringifyContent(
  node: Parent,
  toString: (param: unknown) => string, // TODO weird but works
): string {
  return (node.children as PhrasingContent[])
    .map((item) => toValue(item, toString))
    .join('');
}

// TODO This is really a workaround, and not super reliable
// For now we only support serializing tagName, className and content
// Can we implement the TOC with real JSX nodes instead of html strings later?
function mdxJsxTextElementToHtml(
  element: MdxJsxTextElement,
  toString: (param: unknown) => string, // TODO weird but works
): string {
  const tag = element.name;

  const attributes = element.attributes.filter(
    (child): child is MdxJsxAttribute => child.type === 'mdxJsxAttribute',
  );

  const classAttribute =
    attributes.find((attr) => attr.name === 'className') ??
    attributes.find((attr) => attr.name === 'class');

  const classAttributeString = classAttribute
    ? `class="${escapeHtml(String(classAttribute.value))}"`
    : ``;

  const allAttributes = classAttributeString ? ` ${classAttributeString}` : '';

  const content = stringifyContent(element, toString);

  return `<${tag}${allAttributes}>${content}</${tag}>`;
}

export function toValue(
  node: PhrasingContent | Heading | MdxJsxTextElement,
  toString: (param: unknown) => string, // TODO weird but works
): string {
  switch (node.type) {
    case 'mdxJsxTextElement': {
      return mdxJsxTextElementToHtml(node as MdxJsxTextElement, toString);
    }
    case 'text':
      return escapeHtml(node.value);
    case 'heading':
      return stringifyContent(node, toString);
    case 'inlineCode':
      return `<code>${escapeHtml(node.value)}</code>`;
    case 'emphasis':
      return `<em>${stringifyContent(node, toString)}</em>`;
    case 'strong':
      return `<strong>${stringifyContent(node, toString)}</strong>`;
    case 'delete':
      return `<del>${stringifyContent(node, toString)}</del>`;
    case 'link':
      return stringifyContent(node, toString);
    default:
      return toString(node);
  }
}

export function assetRequireAttributeValue(
  requireString: string,
  hash: string,
): MdxJsxAttributeValueExpression {
  return {
    type: 'mdxJsxAttributeValueExpression',
    value: `require("${requireString}").default${hash && ` + '${hash}'`}`,
    data: {
      estree: {
        type: 'Program',
        body: [
          {
            type: 'ExpressionStatement',
            expression: {
              type: 'BinaryExpression',
              left: {
                type: 'MemberExpression',
                object: {
                  type: 'CallExpression',
                  callee: {
                    type: 'Identifier',
                    name: 'require',
                  },
                  arguments: [
                    {
                      type: 'Literal',
                      value: requireString,
                      raw: `"${requireString}"`,
                    },
                  ],
                  optional: false,
                },
                property: {
                  type: 'Identifier',
                  name: 'default',
                },
                computed: false,
                optional: false,
              },
              operator: '+',
              right: {
                type: 'Literal',
                value: hash,
                raw: `"${hash}"`,
              },
            },
          },
        ],
        sourceType: 'module',
        comments: [],
      },
    },
  };
}
