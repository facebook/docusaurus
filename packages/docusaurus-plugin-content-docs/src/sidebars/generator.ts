/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import _ from 'lodash';
import logger from '@docusaurus/logger';
import {addTrailingSlash} from '@docusaurus/utils-common';
import {createDocsByIdIndex, toCategoryIndexMatcherParam} from '../docs';
import type {
  SidebarItemDoc,
  SidebarItemsGenerator,
  SidebarItemsGeneratorDoc,
  NormalizedSidebarItemCategory,
  NormalizedSidebarItem,
  SidebarItemCategoryLinkConfig,
} from './types';

const BreadcrumbSeparator = '/';

// Just an alias to the make code more explicit
function getLocalDocId(docId: string): string {
  return _.last(docId.split('/'))!;
}

export const CategoryMetadataFilenameBase = '_category_';
export const CategoryMetadataFilenamePattern = '_category_.{json,yml,yaml}';

type WithPosition<T> = T & {
  position?: number;
  /** The source is the file/folder name */
  source?: string;
};

/**
 * A representation of the fs structure. For each object entry:
 * If it's a folder, the key is the directory name, and value is the directory
 * content; If it's a doc file, the key is the doc's source file name, and value
 * is the doc ID
 */
type Dir = {
  [item: string]: Dir | string;
};

// Comment for this feature: https://github.com/facebook/docusaurus/issues/3464#issuecomment-818670449
export const DefaultSidebarItemsGenerator: SidebarItemsGenerator = ({
  numberPrefixParser,
  isCategoryIndex,
  docs: allDocs,
  item: {dirName: autogenDir},
  categoriesMetadata,
}) => {
  const docsById = createDocsByIdIndex(allDocs);
  const findDoc = (docId: string): SidebarItemsGeneratorDoc | undefined =>
    docsById[docId];
  const getDoc = (docId: string): SidebarItemsGeneratorDoc => {
    const doc = findDoc(docId);
    if (!doc) {
      throw new Error(
        `Can't find any doc with ID ${docId}.
Available doc IDs:
- ${Object.keys(docsById).join('\n- ')}`,
      );
    }
    return doc;
  };

  /**
   * Step 1. Extract the docs that are in the autogen dir.
   */
  function getAutogenDocs(): SidebarItemsGeneratorDoc[] {
    function isInAutogeneratedDir(doc: SidebarItemsGeneratorDoc) {
      return (
        // Doc at the root of the autogenerated sidebar dir
        doc.sourceDirName === autogenDir ||
        // Autogen dir is . and doc is in subfolder
        autogenDir === '.' ||
        // Autogen dir is not . and doc is in subfolder
        // "api/myDoc" startsWith "api/" (note "api2/myDoc" is not included)
        doc.sourceDirName.startsWith(addTrailingSlash(autogenDir))
      );
    }
    const docs = allDocs.filter(isInAutogeneratedDir);

    if (docs.length === 0) {
      logger.warn`No docs found in path=${autogenDir}: can't auto-generate a sidebar.`;
    }
    return docs;
  }

  /**
   * Step 2. Turn the linear file list into a tree structure.
   */
  function treeify(docs: SidebarItemsGeneratorDoc[]): Dir {
    // Get the category breadcrumb of a doc (relative to the dir of the
    // autogenerated sidebar item)
    // autogenDir=a/b and docDir=a/b/c/d => returns [c, d]
    // autogenDir=a/b and docDir=a/b => returns []
    // TODO: try to use path.relative()
    function getRelativeBreadcrumb(doc: SidebarItemsGeneratorDoc): string[] {
      return autogenDir === doc.sourceDirName
        ? []
        : doc.sourceDirName
            .replace(addTrailingSlash(autogenDir), '')
            .split(BreadcrumbSeparator);
    }
    const treeRoot: Dir = {};
    docs.forEach((doc) => {
      const breadcrumb = getRelativeBreadcrumb(doc);
      // We walk down the file's path to generate the fs structure
      let currentDir = treeRoot;
      breadcrumb.forEach((dir) => {
        if (typeof currentDir[dir] === 'undefined') {
          currentDir[dir] = {}; // Create new folder.
        }
        currentDir = currentDir[dir] as Dir; // Go into the subdirectory.
      });
      // We've walked through the path. Register the file in this directory.
      currentDir[path.basename(doc.source)] = doc.id;
    });
    return treeRoot;
  }

  /**
   * Step 3. Recursively transform the tree-like structure to sidebar items.
   * (From a record to an array of items, akin to normalizing shorthand)
   */
  function generateSidebar(
    fsModel: Dir,
  ): WithPosition<NormalizedSidebarItem>[] {
    function createDocItem(
      id: string,
      fullPath: string,
      fileName: string,
    ): WithPosition<SidebarItemDoc> {
      const {
        sidebarPosition: position,
        frontMatter: {
          sidebar_label: label,
          sidebar_class_name: className,
          sidebar_custom_props: customProps,
        },
      } = getDoc(id);
      return {
        type: 'doc',
        id,
        position,
        source: fileName,
        // We don't want these fields to magically appear in the generated
        // sidebar
        ...(label !== undefined && {label}),
        ...(className !== undefined && {className}),
        ...(customProps !== undefined && {customProps}),
      };
    }
    function createCategoryItem(
      dir: Dir,
      fullPath: string,
      folderName: string,
    ): WithPosition<NormalizedSidebarItemCategory> {
      const categoryMetadata =
        categoriesMetadata[path.posix.join(autogenDir, fullPath)];
      const allItems = Object.entries(dir).map(([key, content]) =>
        dirToItem(content, key, `${fullPath}/${key}`),
      );

      // Try to match a doc inside the category folder,
      // using the "local id" (myDoc) or "qualified id" (dirName/myDoc)
      function findDocByLocalId(localId: string): SidebarItemDoc | undefined {
        return allItems.find(
          (item): item is SidebarItemDoc =>
            item.type === 'doc' && getLocalDocId(item.id) === localId,
        );
      }

      function findConventionalCategoryDocLink(): SidebarItemDoc | undefined {
        return allItems.find((item): item is SidebarItemDoc => {
          if (item.type !== 'doc') {
            return false;
          }
          const doc = getDoc(item.id);
          return isCategoryIndex(toCategoryIndexMatcherParam(doc));
        });
      }

      // In addition to the ID, this function also retrieves metadata of the
      // linked doc that could be used as fallback values for category metadata
      function getCategoryLinkedDocMetadata():
        | {
            id: string;
            position?: number;
            label?: string;
            customProps?: {[key: string]: unknown};
            className?: string;
          }
        | undefined {
        const link = categoryMetadata?.link;
        if (link !== undefined && link?.type !== 'doc') {
          // If a link is explicitly specified, we won't apply conventions
          return undefined;
        }
        const id = link
          ? findDocByLocalId(link.id)?.id ?? getDoc(link.id).id
          : findConventionalCategoryDocLink()?.id;
        if (!id) {
          return undefined;
        }
        const doc = getDoc(id);
        return {
          id,
          position: doc.sidebarPosition,
          label: doc.frontMatter.sidebar_label ?? doc.title,
          customProps: doc.frontMatter.sidebar_custom_props,
          className: doc.frontMatter.sidebar_class_name,
        };
      }
      const categoryLinkedDoc = getCategoryLinkedDocMetadata();
      const link: SidebarItemCategoryLinkConfig | null | undefined =
        categoryLinkedDoc
          ? {
              type: 'doc',
              id: categoryLinkedDoc.id, // We "remap" a potentially "local id" to a "qualified id"
            }
          : categoryMetadata?.link;
      // If a doc is linked, remove it from the category subItems
      const items = allItems.filter(
        (item) => !(item.type === 'doc' && item.id === categoryLinkedDoc?.id),
      );

      const className =
        categoryMetadata?.className ?? categoryLinkedDoc?.className;
      const customProps =
        categoryMetadata?.customProps ?? categoryLinkedDoc?.customProps;
      const {filename, numberPrefix} = numberPrefixParser(folderName);

      return {
        type: 'category',
        label: categoryMetadata?.label ?? categoryLinkedDoc?.label ?? filename,
        collapsible: categoryMetadata?.collapsible,
        collapsed: categoryMetadata?.collapsed,
        position:
          categoryMetadata?.position ??
          categoryLinkedDoc?.position ??
          numberPrefix,
        source: folderName,
        ...(customProps !== undefined && {customProps}),
        ...(className !== undefined && {className}),
        items,
        ...(categoryMetadata?.description && {
          description: categoryMetadata?.description,
        }),
        ...(link && {link}),
      };
    }
    function dirToItem(
      dir: Dir | string, // The directory item to be transformed.
      itemKey: string, // File/folder name; for categories, it's used to generate the next `relativePath`.
      fullPath: string, // `dir`'s full path relative to the autogen dir.
    ): WithPosition<NormalizedSidebarItem> {
      return typeof dir === 'object'
        ? createCategoryItem(dir, fullPath, itemKey)
        : createDocItem(dir, fullPath, itemKey);
    }
    return Object.entries(fsModel).map(([key, content]) =>
      dirToItem(content, key, key),
    );
  }

  /**
   * Step 4. Recursively sort the categories/docs + remove the "position"
   * attribute from final output. Note: the "position" is only used to sort
   * "inside" a sidebar slice. It is not used to sort across multiple
   * consecutive sidebar slices (i.e. a whole category composed of multiple
   * autogenerated items)
   */
  function sortItems(
    sidebarItems: WithPosition<NormalizedSidebarItem>[],
  ): NormalizedSidebarItem[] {
    const processedSidebarItems = sidebarItems.map((item) => {
      if (item.type === 'category') {
        return {...item, items: sortItems(item.items)};
      }
      return item;
    });
    const sortedSidebarItems = _.sortBy(processedSidebarItems, [
      'position',
      'source',
    ]);
    return sortedSidebarItems.map(({position, source, ...item}) => item);
  }
  // TODO: the whole code is designed for pipeline operator
  const docs = getAutogenDocs();
  const fsModel = treeify(docs);
  const sidebarWithPosition = generateSidebar(fsModel);
  const sortedSidebar = sortItems(sidebarWithPosition);
  return sortedSidebar;
};
