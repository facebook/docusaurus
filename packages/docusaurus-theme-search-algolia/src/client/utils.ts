/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {FacetFilters} from 'algoliasearch/lite';

export function mergeFacetFilters(
  f1: FacetFilters,
  f2: FacetFilters,
): FacetFilters;

export function mergeFacetFilters(
  f1: FacetFilters | undefined,
  f2: FacetFilters | undefined,
): FacetFilters | undefined;

export function mergeFacetFilters(
  f1: FacetFilters | undefined,
  f2: FacetFilters | undefined,
): FacetFilters | undefined {
  if (f1 === undefined) {
    return f2;
  }
  if (f2 === undefined) {
    return f1;
  }

  const normalize = (f: FacetFilters): FacetFilters =>
    typeof f === 'string' ? [f] : f;

  // Historical behavior: we flatten everything
  // TODO I'm pretty sure this is incorrect
  //  see https://www.algolia.com/doc/api-reference/api-parameters/facetFilters/?client=javascript
  //  Note: Algolia is working to provide a reliable facet merging strategy
  //  see https://github.com/facebook/docusaurus/pull/11327#issuecomment-3284742923
  return [...normalize(f1), ...normalize(f2)];
}
