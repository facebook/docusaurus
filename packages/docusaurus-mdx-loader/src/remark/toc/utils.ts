/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {toValue} from '../utils';
import type {Node} from 'unist';
import type {
  MdxjsEsm,
  // @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
} from 'mdast-util-mdx';
import type {TOCHeading, TOCItem, TOCItems, TOCSlice} from './types';
import type {
  Program,
  SpreadElement,
  ImportDeclaration,
  ImportSpecifier,
} from 'estree';

export function getImportDeclarations(program: Program): ImportDeclaration[] {
  return program.body.filter(
    (item): item is ImportDeclaration => item.type === 'ImportDeclaration',
  );
}

export function isMarkdownImport(node: Node): node is ImportDeclaration {
  if (node.type !== 'ImportDeclaration') {
    return false;
  }
  const importPath = (node as ImportDeclaration).source.value;
  return typeof importPath === 'string' && /\.mdx?$/.test(importPath);
}

export function findDefaultImportName(
  importDeclaration: ImportDeclaration,
): string | undefined {
  return importDeclaration.specifiers.find(
    (o: Node) => o.type === 'ImportDefaultSpecifier',
  )?.local.name;
}

export function findNamedImportSpecifier(
  importDeclaration: ImportDeclaration,
  localName: string,
): ImportSpecifier | undefined {
  return importDeclaration?.specifiers.find(
    (specifier): specifier is ImportSpecifier =>
      specifier.type === 'ImportSpecifier' &&
      specifier.local.name === localName,
  );
}

// Before: import Partial from "partial"
// After: import Partial, {toc as __tocPartial} from "partial"
export function addTocSliceImportIfNeeded({
  importDeclaration,
  tocExportName,
  tocSliceImportName,
}: {
  importDeclaration: ImportDeclaration;
  tocExportName: string;
  tocSliceImportName: string;
}): void {
  // We only add the toc slice named import if it doesn't exist already
  if (!findNamedImportSpecifier(importDeclaration, tocSliceImportName)) {
    importDeclaration.specifiers.push({
      type: 'ImportSpecifier',
      imported: {type: 'Identifier', name: tocExportName},
      local: {type: 'Identifier', name: tocSliceImportName},
    });
  }
}

export function isNamedExport(
  node: Node,
  exportName: string,
): node is MdxjsEsm {
  if (node.type !== 'mdxjsEsm') {
    return false;
  }
  const program = (node as MdxjsEsm).data?.estree;
  if (!program) {
    return false;
  }
  if (program.body.length !== 1) {
    return false;
  }
  const exportDeclaration = program.body[0]!;
  if (exportDeclaration.type !== 'ExportNamedDeclaration') {
    return false;
  }
  const variableDeclaration = exportDeclaration.declaration;
  if (variableDeclaration?.type !== 'VariableDeclaration') {
    return false;
  }
  const {id} = variableDeclaration.declarations[0]!;
  if (id.type !== 'Identifier') {
    return false;
  }

  return id.name === exportName;
}

export async function createTOCExportNodeAST({
  tocExportName,
  tocItems,
}: {
  tocExportName: string;
  tocItems: TOCItems;
}): Promise<MdxjsEsm> {
  function createTOCSliceAST(tocSlice: TOCSlice): SpreadElement {
    return {
      type: 'SpreadElement',
      argument: {type: 'Identifier', name: tocSlice.importName},
    };
  }

  async function createTOCHeadingAST({heading}: TOCHeading) {
    const {toString} = await import('mdast-util-to-string');
    const {valueToEstree} = await import('estree-util-value-to-estree');
    const value: TOCItem = {
      value: toValue(heading, toString),
      id: heading.data!.id!,
      level: heading.depth,
    };
    return valueToEstree(value);
  }

  async function createTOCItemAST(tocItem: TOCItems[number]) {
    switch (tocItem.type) {
      case 'slice':
        return createTOCSliceAST(tocItem);
      case 'heading':
        return createTOCHeadingAST(tocItem);
      default: {
        throw new Error(`unexpected toc item type`);
      }
    }
  }

  return {
    type: 'mdxjsEsm',
    value: '', // See https://github.com/facebook/docusaurus/pull/9684#discussion_r1457595181
    data: {
      estree: {
        type: 'Program',
        body: [
          {
            type: 'ExportNamedDeclaration',
            declaration: {
              type: 'VariableDeclaration',
              declarations: [
                {
                  type: 'VariableDeclarator',
                  id: {
                    type: 'Identifier',
                    name: tocExportName,
                  },
                  init: {
                    type: 'ArrayExpression',
                    elements: await Promise.all(tocItems.map(createTOCItemAST)),
                  },
                },
              ],
              kind: 'const',
            },
            specifiers: [],
            source: null,
          },
        ],
        sourceType: 'module',
      },
    },
  };
}
