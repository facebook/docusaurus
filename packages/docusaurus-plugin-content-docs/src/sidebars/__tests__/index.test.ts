/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {jest} from '@jest/globals';
import path from 'path';
import {createSlugger} from '@docusaurus/utils';
import {loadSidebars, DisabledSidebars} from '../index';
import {DefaultSidebarItemsGenerator} from '../generator';
import type {SidebarProcessorParams} from '../types';
import type {
  DocMetadata,
  VersionMetadata,
} from '@docusaurus/plugin-content-docs';

describe('loadSidebars', () => {
  const fixtureDir = path.join(__dirname, '__fixtures__', 'sidebars');
  const params: SidebarProcessorParams = {
    sidebarItemsGenerator: DefaultSidebarItemsGenerator,
    numberPrefixParser: (filename) => ({filename}),
    docs: [
      {
        source: '@site/docs/foo/bar.md',
        sourceDirName: 'foo',
        id: 'bar',
        frontMatter: {},
      },
    ] as DocMetadata[],
    drafts: [],
    version: {
      path: 'version',
      contentPath: path.join(fixtureDir, 'docs'),
      contentPathLocalized: path.join(fixtureDir, 'docs'),
    } as VersionMetadata,
    categoryLabelSlugger: {slug: (v) => v},
    sidebarOptions: {sidebarCollapsed: true, sidebarCollapsible: true},
  };
  it('sidebars with known sidebar item type', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars.json');
    const result = await loadSidebars(sidebarPath, params);
    expect(result).toMatchSnapshot();
  });

  it('sidebars with some draft items', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars.json');
    const paramsWithDrafts: SidebarProcessorParams = {
      ...params,
      drafts: [{id: 'foo/baz'} as DocMetadata, {id: 'hello'} as DocMetadata],
    };
    const result = await loadSidebars(sidebarPath, paramsWithDrafts);
    expect(result).toMatchSnapshot();
  });

  it('sidebars with deep level of category', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars-category.js');
    const result = await loadSidebars(sidebarPath, params);
    expect(result).toMatchSnapshot();
  });

  it('sidebars shorthand and longhand lead to exact same sidebar', async () => {
    const sidebarPath1 = path.join(fixtureDir, 'sidebars-category.js');
    const sidebarPath2 = path.join(
      fixtureDir,
      'sidebars-category-shorthand.js',
    );
    const sidebar1 = await loadSidebars(sidebarPath1, params);
    const sidebar2 = await loadSidebars(sidebarPath2, params);
    expect(sidebar1).toEqual(sidebar2);
  });

  it('sidebars with category but category.items is not an array', async () => {
    const sidebarPath = path.join(
      fixtureDir,
      'sidebars-category-wrong-items.json',
    );
    await expect(() =>
      loadSidebars(sidebarPath, params),
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"Invalid sidebar items collection \`"doc1"\` in \`items\` of the category Category Label: it must either be an array of sidebar items or a shorthand notation (which doesn't contain a \`type\` property). See https://docusaurus.io/docs/sidebar/items for all valid syntaxes."`,
    );
  });

  it('sidebars with first level not a category', async () => {
    const sidebarPath = path.join(
      fixtureDir,
      'sidebars-first-level-not-category.js',
    );
    const result = await loadSidebars(sidebarPath, params);
    expect(result).toMatchSnapshot();
  });

  it('sidebars link', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars-link.json');
    const result = await loadSidebars(sidebarPath, params);
    expect(result).toMatchSnapshot();
  });

  it('nonexistent path', async () => {
    await expect(loadSidebars('bad/path', params)).resolves.toEqual(
      DisabledSidebars,
    );
  });

  it('undefined path', async () => {
    await expect(loadSidebars(undefined, params)).resolves.toMatchSnapshot();
  });

  it('literal false path', async () => {
    await expect(loadSidebars(false, params)).resolves.toEqual(
      DisabledSidebars,
    );
  });

  it('sidebars with category.collapsed property', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars-collapsed.json');
    const result = await loadSidebars(sidebarPath, params);
    expect(result).toMatchSnapshot();
  });

  it('sidebars with category.collapsed property at first level', async () => {
    const sidebarPath = path.join(
      fixtureDir,
      'sidebars-collapsed-first-level.json',
    );
    const result = await loadSidebars(sidebarPath, params);
    expect(result).toMatchSnapshot();
  });

  it('loads sidebars with index-only categories', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars-category-index.json');
    const result = await loadSidebars(sidebarPath, {
      ...params,
      docs: [
        {
          id: 'tutorials/tutorial-basics',
          source: '@site/docs/tutorials/tutorial-basics/index.md',
          sourceDirName: 'tutorials/tutorial-basics',
          frontMatter: {},
        },
      ] as DocMetadata[],
    });
    expect(result).toMatchSnapshot();
  });

  it('loads sidebars with interspersed draft items', async () => {
    const sidebarPath = path.join(fixtureDir, 'sidebars-drafts.json');
    const result = await loadSidebars(sidebarPath, {
      ...params,
      drafts: [{id: 'draft1'}, {id: 'draft2'}, {id: 'draft3'}] as DocMetadata[],
      categoryLabelSlugger: createSlugger(),
    });
    expect(result).toMatchSnapshot();
  });

  it('duplicate category metadata files', async () => {
    const sidebarPath = path.join(
      fixtureDir,
      'sidebars-collapsed-first-level.json',
    );
    const consoleWarnMock = jest
      .spyOn(console, 'warn')
      .mockImplementation(() => {});
    const consoleErrorMock = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});
    await expect(() =>
      loadSidebars(sidebarPath, {
        ...params,
        version: {
          contentPath: path.join(fixtureDir, 'invalid-docs'),
          contentPathLocalized: path.join(fixtureDir, 'invalid-docs'),
        } as VersionMetadata,
      }),
    ).rejects.toThrowErrorMatchingInlineSnapshot(`""foo" is not allowed"`);
    expect(consoleWarnMock).toHaveBeenCalledWith(
      expect.stringMatching(
        /.*\[WARNING\].* There are more than one category metadata files for .*foo.*: foo\/_category_.json, foo\/_category_.yml. The behavior is undetermined./,
      ),
    );
    expect(consoleErrorMock).toHaveBeenCalledWith(
      expect.stringMatching(
        /.*\[ERROR\].* The docs sidebar category metadata file .*foo\/_category_.json.* looks invalid!/,
      ),
    );
  });
});
