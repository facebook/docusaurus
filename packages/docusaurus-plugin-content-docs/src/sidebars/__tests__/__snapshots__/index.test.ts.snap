// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadSidebars loads sidebars with index-only categories 1`] = `
{
  "docs": [
    {
      "collapsed": true,
      "collapsible": true,
      "items": [
        {
          "id": "tutorials/tutorial-basics",
          "label": "tutorial-basics",
          "type": "doc",
        },
      ],
      "label": "Tutorials",
      "link": undefined,
      "type": "category",
    },
    {
      "id": "tutorials/tutorial-basics",
      "label": "index-only",
      "type": "doc",
    },
  ],
}
`;

exports[`loadSidebars loads sidebars with interspersed draft items 1`] = `
{
  "sidebar": [
    {
      "id": "not-draft",
      "label": "index not draft",
      "type": "doc",
    },
    {
      "collapsed": true,
      "collapsible": true,
      "items": [
        {
          "id": "not-draft",
          "type": "doc",
        },
      ],
      "label": "subitem not draft",
      "link": undefined,
      "type": "category",
    },
  ],
}
`;

exports[`loadSidebars sidebars link 1`] = `
{
  "docs": [
    {
      "collapsed": true,
      "collapsible": true,
      "items": [
        {
          "href": "https://github.com",
          "label": "category",
          "type": "link",
        },
      ],
      "label": "Test",
      "link": undefined,
      "type": "category",
    },
  ],
}
`;

exports[`loadSidebars sidebars with category.collapsed property 1`] = `
{
  "docs": [
    {
      "collapsed": true,
      "collapsible": true,
      "items": [
        {
          "collapsed": false,
          "collapsible": true,
          "items": [
            {
              "id": "doc1",
              "type": "doc",
            },
          ],
          "label": "Introduction",
          "link": undefined,
          "type": "category",
        },
      ],
      "label": "Test",
      "link": undefined,
      "type": "category",
    },
    {
      "collapsed": true,
      "collapsible": true,
      "items": [
        {
          "collapsed": false,
          "collapsible": true,
          "items": [
            {
              "id": "doc2",
              "type": "doc",
            },
          ],
          "label": "Powering MDX",
          "link": undefined,
          "type": "category",
        },
      ],
      "label": "Reference",
      "link": undefined,
      "type": "category",
    },
  ],
}
`;

exports[`loadSidebars sidebars with category.collapsed property at first level 1`] = `
{
  "docs": [
    {
      "collapsed": false,
      "collapsible": true,
      "items": [
        {
          "id": "doc1",
          "type": "doc",
        },
      ],
      "label": "Introduction",
      "link": undefined,
      "type": "category",
    },
    {
      "collapsed": false,
      "collapsible": true,
      "items": [
        {
          "id": "doc2",
          "type": "doc",
        },
      ],
      "label": "Powering MDX",
      "link": undefined,
      "type": "category",
    },
  ],
}
`;

exports[`loadSidebars sidebars with deep level of category 1`] = `
{
  "docs": [
    {
      "collapsed": true,
      "collapsible": true,
      "items": [
        {
          "id": "a",
          "type": "doc",
        },
        {
          "collapsed": true,
          "collapsible": true,
          "items": [
            {
              "collapsed": true,
              "collapsible": true,
              "items": [
                {
                  "id": "c",
                  "type": "doc",
                },
                {
                  "collapsed": true,
                  "collapsible": true,
                  "items": [
                    {
                      "id": "d",
                      "type": "doc",
                    },
                    {
                      "collapsed": true,
                      "collapsible": true,
                      "items": [
                        {
                          "id": "e",
                          "type": "doc",
                        },
                      ],
                      "label": "deeper more more",
                      "link": undefined,
                      "type": "category",
                    },
                  ],
                  "label": "level 4",
                  "link": undefined,
                  "type": "category",
                },
              ],
              "label": "level 3",
              "link": undefined,
              "type": "category",
            },
            {
              "id": "f",
              "type": "doc",
            },
          ],
          "label": "level 2",
          "link": undefined,
          "type": "category",
        },
      ],
      "label": "level 1",
      "link": undefined,
      "type": "category",
    },
  ],
}
`;

exports[`loadSidebars sidebars with first level not a category 1`] = `
{
  "docs": [
    {
      "collapsed": true,
      "collapsible": true,
      "items": [
        {
          "id": "greeting",
          "type": "doc",
        },
      ],
      "label": "Getting Started",
      "link": undefined,
      "type": "category",
    },
    {
      "id": "api",
      "type": "doc",
    },
  ],
}
`;

exports[`loadSidebars sidebars with known sidebar item type 1`] = `
{
  "docs": [
    {
      "collapsed": true,
      "collapsible": true,
      "items": [
        {
          "id": "foo/bar",
          "type": "doc",
        },
        {
          "id": "foo/baz",
          "type": "doc",
        },
        {
          "href": "https://github.com",
          "label": "GitHub",
          "type": "link",
        },
        {
          "id": "hello",
          "type": "ref",
        },
      ],
      "label": "Test",
      "link": undefined,
      "type": "category",
    },
    {
      "collapsed": true,
      "collapsible": true,
      "items": [
        {
          "id": "hello",
          "type": "doc",
        },
      ],
      "label": "Guides",
      "link": undefined,
      "type": "category",
    },
  ],
}
`;

exports[`loadSidebars sidebars with some draft items 1`] = `
{
  "docs": [
    {
      "collapsed": true,
      "collapsible": true,
      "items": [
        {
          "id": "foo/bar",
          "type": "doc",
        },
        {
          "href": "https://github.com",
          "label": "GitHub",
          "type": "link",
        },
      ],
      "label": "Test",
      "link": undefined,
      "type": "category",
    },
  ],
}
`;

exports[`loadSidebars undefined path 1`] = `
{
  "defaultSidebar": [
    {
      "collapsed": true,
      "collapsible": true,
      "items": [
        {
          "id": "bar",
          "type": "doc",
        },
      ],
      "label": "foo",
      "link": undefined,
      "type": "category",
    },
  ],
}
`;
